[Wong Zheng Jie] iP

KrashKart opened the pull request:
# Jackson
> At any one time, language is a kaleidoscope of genres, styles and dialects - [David Crystal](https://quotefancy.com/quote/1151562/David-Crystal-At-any-one-time-language-is-a-kaleidoscope-of-styles-genres-and-dialects)

Jackson is not :no_entry_sign: *any* ordinary chat-bot and task tracker. Devoid of posh and proper English, Jackson embraces the imperfection of **Singlish** in a uniquely ~~rabakz~~ Singaporean way :metal:.

## Installing Jackson
Before you *lim kopi* :coffee: (drink coffee) with Jackson, ensure that you:
1. Download the jar file [here](https://github.com/KrashKart/ip/releases/latest)
2. Run the jar file
3. Start adding your tasks!

## Why Jackson?
- Trust Jackson to track your tasks from events to deadlines :clipboard: 
- Simply `mark` or `unmark` your tasks to track completion :white_check_mark:
- Forgot a specific task? Jackson comes with in-built `find` functionalities :mag:
- Up for the challenge? We dare you to find a ~~cool easter egg~~ within Jackson :shushing_face:

## Up and Coming Features
- [x] Add save file saving and reading features for continuity
- [x] Add standardised date formatting and storage
- [x] Add GUI
- [x] Update easter egg

## Exploring with Jackson
If you would like to experiment or integrate Jackson into your own products, provided below is the `main` code:
```java
public static void main(String[] args) {
    // create instance of Jackson and run main loop
    Jackson jackson = new Jackson(EXPECTED_SIZE, PATH);
    jackson.run();
}
```
where `EXPECTED_SIZE` and `PATH` are the expected task list sizes (can be violated dynamically) and the path for the save file (can be set dynamically).

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

aureliony committed:
build.gradle: Prevent generating a second JAR file

In build.gradle, the dependencies on distZip and/or distTar causes
the shadowJar task to generate a second JAR file for which the
mainClass.set("seedu.duke.Duke") does not take effect.
Hence, this additional JAR file cannot be run.
For this product, there is no need to generate a second JAR file
to begin with.

Let's remove this dependency from the build.gradle to prevent the
shadowJar task from generating the extra JAR file.

KrashKart committed:
changed Duke.java to Jackson.java
changed greeting

KrashKart committed:
added conversation loop

KrashKart committed:
added add and list functions, segregated add and list into separate global methods

KrashKart committed:
changed attribute names

KrashKart committed:
Added Task class file

KrashKart committed:
Added marking and unmarking functionality into chatbot loop

KrashKart committed:
Added Deadline, Event and Todo class files, changed main loop from switch statement to if else (because it's easier for me), established regex parsing of input and tidied up code

KrashKart committed:
added automated UI testing files

KrashKart committed:
Added custom exception files, modified main loop to utilise REGEX (for neater and shorter code)

KrashKart committed:
Added delete method and improved Exceptions to include: invalid index (0) and invalid index (above list size)

KrashKart committed:
Integrated "bye" detection into if-else loop, added secret function

KrashKart committed:
Added encryption for secret text file

KrashKart committed:
Grouped Exceptions under JacksonException, removed redundant InvalidIndexException

KrashKart committed:
Added enums for Response type checking

KrashKart committed:
Added JacksonException as the parent Exception class for Jackson-specific exceptions, standardized all JacksonExceptions to take in a String msg during construction

KrashKart committed:
Created Response, TaskList and Parser class to handle Action type and Matcher, task list amendments and input parsing respectively

KrashKart committed:
Updated UI testing files for the latest version of Jackson

KrashKart committed:
Added javadocs to existing methods/variables and comments describing main loop

KrashKart committed:
Modify Javadocs description

Javadocs description does not adhere to coding standards.

Javadocs have to conform to standards according to course demands.

Let's rewrite Javadocs to adhere to standards in terms of description.

KrashKart committed:
TaskList: Remove showList, add overloaded toString

TaskList currently uses showList to show the task list

This is not generalisable as we now have to be able to show and save
tasks to file.

Let's remove showList and implement a more flexible toString
implementation so it can be used for both "list" and "save" commands.

KrashKart committed:
TaskList: Add save method

TaskList currently cannot save the list to a text file after every
modification.

Upon saving, users can retrieve task list history if needed.

Let's add the save function to write the task list history into a
text file in real-time.

KrashKart committed:
gitignore: Add data.txt file

gitignore currently doesn't ignore the task list save file data.txt.

This save file is unnecessary to push to git as it will clutter
the repo.

Let's add data.txt to the gitignore.

KrashKart committed:
Add save and load functionalities

Currently, Jackson does not save the task list after changes and
cannot load a previously saved task list.

This means every time we start Jackson, we have to waste time
re-entering all the tasks, which is cumbersome and not functional.

Let's add this save and load function such that the tasks can be
written and read from a save file data.txt to ensure continuity of
memory across Jackson runs.
* Modify TaskList to include save and load methods
* Add regex to read tasks from the save text file more easily
* Modify TaskList modification methods (addList, deleteList, etc)
to save after each method call
* Add Exception checking for IOExceptions
* Remove toString method as it might be redundant and reinstate
showList

KrashKart committed:
Rename secret command

Secret command is too obvious

Thus, users can guess functionality based on the command itself

Let's change the command to make it less obvious

KrashKart committed:
Merge branch 'branch-Level-7'

KrashKart committed:
Temporal: Add new DateTime/Date class

Previously, there was no restriction on dates for Event and
Deadline classes.

Thus, there was no standardized format for users to input their
from, to and by dates.

Let's do a few things here:
* Create Temporal class to handle DateTime/Date information storage
* Create DateTimeFormatters to output datetime accordingly
* Modify Event and Deadline classes to contain Temporal class
objects
* Modify Temporal class to be able to read saved dates and
datetimes (which have already been converted to output datetime
format)

KrashKart committed:
Remove unused imports

Get rid of unused imports in Jackson and Parser, as well as
modifying regex patterns to comply with accepted datetime input
format

KrashKart committed:
Merge branch 'branch-Level-8'

KrashKart committed:
More OOP: Add new class files

Previously, Jackson.java handled most of the functionality for the
chatbot (eg. User Interaction and Save File input-output), which
made the file messy and convoluted.

In addition, TaskList.java also had untidy code due to placing too
much functionality in one file. Overall code was not segmented well
in accordance to OOP principles.

Let's do a few changes:
* Create new Ui class to handle printing output and exceptions
* Transfer printing responsibility from Jackson to Ui
* Create new Storage class to handle save file input-output
* Transfer read-write duties from TaskList to Storage
* Change Jackson from static to instance class
* Separate main and run methods in Jackson, so now main creates
instance of Jackson and runs the main loop instead
* Other code standard changes (naming, etc)

KrashKart committed:
Javadocs: Add java docs to Temporal and newly created classes

Add javadocs for methods and variables in Temporal, Storage, Ui,
Jackson and TaskList

KrashKart committed:
Merge branch 'A-MoreOOP'

KrashKart committed:
Package: Separate files into packages

Files were disorganised without partitions to organise them.

Let's separate all of them into packages for better organisation.

KrashKart committed:
Revert "Package: Separate files into packages"

This reverts commit 3235df536be4b91ff1c0cafdcd1dae9766ce865b.

KrashKart committed:
Reapply "Package: Separate files into packages"

This reverts commit e974cf947cad16d50b5606e39310842bf3154a63.

KrashKart committed:
Packages: Separate files into packages

Files were disorganised previously.

If not organised properly, files can be hard to find and import.

Let's separate all the files into respective packages.

KrashKart committed:
Revert "Packages: Separate files into packages"

This reverts commit 9154d3d42805f97cc1d76a99146155f41a072f6d.

KrashKart committed:
Reapply "Packages: Separate files into packages"

This reverts commit 3c2ed227bc1666ad86c83ac9823e5a54fc8a1775.

KrashKart committed:
Comments: Change comments to be clearer

KrashKart committed:
Merge branch 'A-Packages'

KrashKart committed:
Merge branch 'add-gradle-support'

KrashKart committed:
Gradle: Add Gradle to project

Previously, Grade was not present.

Thus, project tasks could not be automated.

Let's add Gradle support.

KrashKart committed:
Checkstyle: Change code to conform to checkstyle

Run checkstyle on gradle and correct all checkstyle issues in all
files

KrashKart committed:
Merge branch 'branch-A-Gradle'

# Conflicts:
#	build.gradle
#	gradle/wrapper/gradle-wrapper.jar
#	gradle/wrapper/gradle-wrapper.properties
#	gradlew
#	gradlew.bat

KrashKart committed:
Resolve merge conflicts

Due to inconsistent merging with branch-A-gradle, resolve merge
conflicts.

KrashKart committed:
Resolve more merge conflicts

Follow up from previous merge conflict resolution

KrashKart committed:
Parser: Modify exception thrown in parse and add unit tests

Previously, no unit tests existed for any class in Jackson

This is a problem since we cannot automate tests for Jackson
classes and weed out bugs.

Let's:
* Create unit tests for the Parser class (ParserTest)
* Modify Parser to only throw SyntaxException (for standardization)

KrashKart committed:
Ui: Add unit tests

Let's add unit tests to Ui class so we can automate testing better.
This unit tests extensively tests the printFormatGuide method in
Ui

KrashKart committed:
Checkstyle: Rewrite UiTest and ParserTest

Let's
* Run checkstyle via Gradle to see checkstyle errors
* Correct them in both UiTest and ParserTest

KrashKart committed:
Merge branch 'branch-A-JUnit'

KrashKart committed:
ShadowJar: Configure build.gradle for shadowJar conversion

Previously, no jar conversion was available.

This is a problem if jar conversion is required.

Let's modify build.gradle to
* include dependencies for shadowJar
* Modify shadowJar archiveFileName to jackson.jar
* run shadowJar via Gradle to get jackson.jar

jackson.jar will be gitignored

KrashKart committed:
Merge branch 'branch-A-Jar'

KrashKart committed:
Javadoc: Add comments and javadocs to files

Add comments for further elaboration for newer files from
A-MoreOOP tag onwards

KrashKart committed:
Coding Standard: Change code to comply with standards

Placeholder commit for the branch so I can merge it.
Ensure that all files adhere to the coding standards

KrashKart committed:
Find: Add find functionality

Previously, chatbot could not find tasks by searching for keywords.

This means users had to find their desired tasks from a potentially
very long list of tasks, which is cumbersome.

Let's change a few things:
* Add find functionality to Parser, Jackson, Ui and TaskList
* Add find action type to Actions
* Modify unit tests to include testing for find function

KrashKart committed:
Merge branch 'branch-A-JavaDoc'

KrashKart committed:
Merge branch 'branch-A-CodingStandard'

KrashKart committed:
Merge branch 'branch-Level-9'

KrashKart committed:
Clean up imports and add unit test functionality

* Add before each and after each wrappers for junit tests.
* Remove unused imports.

KrashKart committed:
Fix Javadocs punctuation

Add full stop to every line in Javadocs.

KrashKart committed:
CheckStyle: Add Javadocs

Add Javadocs to conform to CheckStyle style requirements.

Specifically, add Javadocs to Jackson.main

KrashKart committed:
Merge branch 'branch-A-CheckStyle'

KrashKart committed:
Varargs: Overload constructor for TaskList

Previously, no constructor for TaskList existed that accepted
Varargs of Tasks.

This renders TaskList inflexible in its construction.

Let's add an overloaded varargs constructor for TaskList

KrashKart committed:
Merge branch 'branch-A-Varargs'

KrashKart committed:
Add tutorial files for JavaFX

KrashKart committed:
JavaFX: Add JavaFX GUI for Jackson

Previously, only text GUI existed for Jackson.

This meant User Experience was not as immersive.

Let's do a few things:
- Add JavaFX GUI drawing, loading and controller files
- Redirect IO from Jackson to JavaFX GUI
- Reformat most functions to return String (JavaFX compatibility)
- Add GUI Quality of Life and Design changes

KrashKart committed:
Improve graphics in GUI

Let's
* Reshape userInput and sendButton to be longer and rounder
respectively
* Add grey pane behind userInput and sendButton to emulate normal
messaging apps
* Add window icon and name
* Add exit timer of 5 seconds and print file save status and exit
message after user chooses to exit

KrashKart committed:
Rectify CheckStyle changes and add enums

Let's
* Make changes to any and all CheckStyle violations to ensure
consistent code quality.
* Add enums support for Jackson DialogBox style logic to ensure
code safety.

KrashKart committed:
Merge branch 'branch-Level-10'

# Conflicts:
#	src/main/java/jackson/Jackson.java

KrashKart committed:
Edit build.gradle for more concise jar name

btbrandon commented on src/main/java/jackson/utils/Parser.java:
```diff
@@ -0,0 +1,106 @@
+ */
+public class Parser {
+
+    /* Bunch of regex patterns for user input checking */
```
Perhaps you can elaborate more on what regex does?

KrashKart replied:
fair enough! will update the code to explain a bit more about regex. thanks for the review!

btbrandon commented on src/main/java/jackson/Jackson.java:
```diff
@@ -0,0 +1,225 @@
+    /**
+     * Runs the main loop of the chatbot.
+     */
+    public String getResponse(String input) {
```
code looks very structured and well-designed!

btbrandon commented on src/main/java/jackson/exceptions/UnsupportedException.java:
```diff
@@ -0,0 +1,10 @@
+/**
+ * Thrown when unrecognised command and input is given.
+ */
+public class UnsupportedException extends JacksonException {
```
could have a clearer name for describing what exactly is unsupported

btbrandon commented on src/main/java/jackson/utils/Parser.java:
```diff
@@ -0,0 +1,106 @@
+        Matcher m = null;
+
+        /*
+            Generally:
```
nice comment! makes it clearer

btbrandon commented on src/main/java/jackson/utils/Temporal.java:
```diff
@@ -0,0 +1,56 @@
+public class Temporal {
+    // DateTimeFormatters for reading (first 2) from save file and user input
+    // and writing (last 2) to save file
+    private static final DateTimeFormatter dateOnly = new DateTimeFormatterBuilder()
```
naming conventions are specific and clearly differentiates the meaning good job!

vicelyav commented on src/main/java/jackson/Jackson.java:
```diff
@@ -0,0 +1,225 @@
+                output = temp.toString().strip();
+
+                // decode from base64 to utf8
+                byte[] decoded = Base64.getDecoder().decode(output);
```
Maybe this one could add more description to the variable on what is being decoded?

KrashKart replied:
thanks for the review! however, i wanted to make this an easter egg (albeit poorly hidden) as much as possible, so had to be vague with the names here

vicelyav commented on src/main/java/jackson/Jackson.java:
```diff
@@ -0,0 +1,225 @@
+     * If secret file not found, handles exception and prints error message.
+     * @return String response.
+     */
+    public String readSecret() {
```
Maybe we could rename it as "readSecretFile()" so we know what is secret.

vicelyav commented on src/main/java/jackson/Jackson.java:
```diff
@@ -0,0 +1,225 @@
+     */
+    public String readSecret() {
+        // get string builder to read line by line
+        StringBuilder temp = new StringBuilder();
```
Maybe could rename this variable a bit to show what temp String we're using it for?

vicelyav commented on src/main/java/jackson/utils/Ui.java:
```diff
@@ -0,0 +1,213 @@
+     * @param taskList {@code TaskList} object task is to be deleted from.
+     * @return String response.
+     */
+    public String printDeleteList(Task task, TaskList taskList) {
```
Maybe we could rename this to show that we're printing after deleting something on the list? Right now when I first read it, I thought it is print, then delete list.

KrashKart replied:
fair point! what i had in mind was that the command given was "deleteList", thus to print that command the method should be called "printDeleteList", but i get your point. will change it and thanks for the review :+1:

vicelyav commented on src/main/java/jackson/utils/Ui.java:
```diff
@@ -0,0 +1,213 @@
+     * @param task {@code Task} object to be marked.
+     * @return String response.
+     */
+    public String printMark(Task task) {
```
Maybe we could rename this to show that we are printing Marked Task instead of printing Mark (could be misinterpreted)?

vicelyav reviewed (commented):
Overall good job on writing a very neat code. I think for the methods, some of them can be renamed/edited to show clearer what the method is doing. Cheers!

KrashKart committed:
Update repository

KrashKart committed:
CodeQuality: Improve code quality

Variable names and comments are not clear enough.

Clearer documentation and more intuitive naming conventions enable
smoother debugging and easier understanding of code.

Let's,
* add short explanation for regex checking in Parser.java
* change method names to be more concise and clear

KrashKart committed:
Add Assertions

No assertions present in the code.

Invariants in the program cannot be tested before chatbot is run,
meaning higher chance of unchecked exceptions.

Let's use Java's assert to check for invariants in the Storage
class like whether the save file is created after saving or if it
exists before loading from it.

KrashKart committed:
Add Stream to read secret file

No streams present in the current program.

Without the versatility of streams, errors and null values cannot
be handled efficiently and lazily.

Let's add streams when reading the secret file in the Jackson.java
file.

This is done to leverage the Optional class in Java for lazy
loading, in case the secret file text changes and becomes long and
more inefficient to load.

KrashKart committed:
Merge pull request #3 from KrashKart/branch-A-Assertions

Add Assertions

KrashKart committed:
Merge branch 'master' of https://github.com/KrashKart/ip

KrashKart committed:
Merge branch 'master' into branch-A-CodeQuality

KrashKart committed:
Merge branch 'master' into branch-A-Streams

KrashKart committed:
Merge pull request #1 from KrashKart/branch-A-CodeQuality

CodeQuality: Improve code quality

KrashKart committed:
Merge branch 'master' into branch-A-Streams

KrashKart committed:
Merge pull request #2 from KrashKart/branch-A-Streams

Add Stream to read secret file

KrashKart committed:
Add CI/CD pipeline

No CI/CD pipeline existed.

Development relies mainly on pushing and pulling without automation
to assist, leading to inconsistent updates and performance.

Let's add the gradle.yml file to facilitate CI into Jackson.

KrashKart committed:
Merge branch 'branch-A-CI'

KrashKart committed:
Update permissions to write to Github Actions

KrashKart committed:
Merge branch 'branch-A-CI'

KrashKart committed:
Test for Github Workflow

KrashKart committed:
Merge branch 'branch-A-CI'

KrashKart committed:
Remove unnecessary workflow

KrashKart committed:
Merge branch 'branch-A-CI'

KrashKart committed:
Sort: Integrate sort functionality

Users were not able to sort tasks according to the task attributes.

This lack of organisation can make it hard to organise tasks in
a neat way according to user preference, which defeats the point of
the chatbot.

Let's,
* create a new sort method in TaskList
* make CustomDateTime comparable to facilitate comparison
* add new sort command to Parser, Ui and Jackson
* add Javadocs
* do minor spring cleaning of code comments and existing Javadocs

Directly incorporating comparison functionality into the classes
involved is much better than simple text comparison of the classes
as it is more airtight and reliable.

KrashKart committed:
Merge branch 'branch-C-Sort'

KrashKart committed:
Add data.txt to .gitignore

KrashKart committed:
Fix bugs in sorting methods

Particularly in TaskList and in the Parser.
Additionally fixed gitignore to ignore the correct data.txt.

KrashKart committed:
Javadocs: Update Javadocs

Javadocs are not updated and contain inaccurate and misleading
information.

This makes debugging harder and new readers will not be able to
easily grasp how the code works.

Let's update the Javadocs to include newly added functions in
Jackson and CustomDateTime so users can understand both files
better.

KrashKart committed:
A-BetterGui: Improve GUI

The GUI is minimal and simple with flat colours and no help
command.

The presentation of the textbot can seem mundane and clunky.
Additionally, lack of help command means users cannot seek usage
help for the commands.

Let's,
* revamp DialogBox to include color gradients and defined borders
* integrate a help command into the program
* revamp printFormatGuide in Ui.java for more detailed explanation
* modify whitespaces in accordance with coding standards and
checkStyle

KrashKart committed:
A-Personality: Add personality quirk to Jackson

Jackson gives standard responses and behaves like a normal chatbot.

This can seem mundane and uninteresting to users and they may not
enjoy or remain engaged with it.

Let's add a "Singlish" style response that Jackson can use to
add flavour to the program.

This is a placeholder commit as I have already integrated this
personality many commits ago.

KrashKart committed:
A-MoreErrorHandling: Improve error handling

Jackson does not have the capability to detect conflicting task
names before adding the tasks.

The possibility of a name conflict impeding code progress later on
is high and minimising this risk is optimal for now.

Let's,
* create DuplicatedTaskException to handle such conflicts
* modify addTask in TaskList.java to check for name conflicts first
* modify Javadocs to adhere to coding standard and checkStyle

KrashKart committed:
A-MoreErrorHandling: Improve error handling part 2

Jackson does not have the capability to detect conflicting task
names before adding the tasks.

The possibility of a name conflict impeding code progress later on
is high and minimising this risk is optimal for now.

Let's,
* create DuplicatedTaskException to handle such conflicts
* modify addTask in TaskList.java to check for name conflicts first
* modify Javadocs to adhere to coding standard and checkStyle

I forgot to stage the changes so this commit will be an extension
to the previous one.

KrashKart committed:
Merge branch 'branch-A-BetterGui'

KrashKart committed:
Merge branch 'branch-A-MoreErrorHandling'

KrashKart committed:
Merge branch 'branch-A-Personality'

KrashKart committed:
Merge branch 'master' into branch-A-MoreTesting

KrashKart committed:
A-MoreTesting: Add new tests

Unit tests for TaskList is not present. Also, due to the transition
to the JavaFX GUI and other output updates, UiTest.java contains
outdated JUnit tests.

This slows down development and inhibits unit testing to ensure
code quality.

Let's,
* update UiTest.java
* create unit tests for TaskListTest.java to test task list
functionality

KrashKart committed:
Merge branch 'branch-A-MoreTesting'

KrashKart committed:
Merge branch 'master' into branch-A-AiAssisted

KrashKart committed:
A-AiAssisted: Improve code using ChatGPT

Tasks were unable to switch easily between DateTime formats.

This meant incompatibility and inversatility.

Let's use ChatGPT to assist with the formulation of a DateTime
formatter.

Since I had used ChatGPT to assist with said DateTime formatter
in CustomDateTime.java many commits ago, this is a placeholder
commit.

KrashKart committed:
Merge branch 'branch-A-AiAssisted'

KrashKart committed:
A-UserGuide: Add user guide

KrashKart committed:
Merge branch 'branch-A-UserGuide'

KrashKart committed:
Add Ui.png and edit gitignore

KrashKart committed:
Update README.md

KrashKart committed:
Update README.md part 2

Add /

KrashKart committed:
Update README.md

Reposition image

KrashKart committed:
Update README.md

I dislike markdown

KrashKart committed:
Fix ordering and tidy code

KrashKart committed:
Merge branch 'master' of https://github.com/KrashKart/ip

KrashKart committed:
Edit gitignore and simplify comparison for CustomDateTime

KrashKart committed:
Remove data.txt from repo

This file is not necessary for the repo as it is a local save used
to test code. This can be ignored.

KrashKart committed:
Improve code standard and quality

KrashKart committed:
Final release

KrashKart committed:
Update gitignore

KrashKart committed:
Update README.md

KrashKart committed:
Update README.md

KrashKart committed:
Remove unused variables

KrashKart committed:
Improve code quality
